python/ray/rllib/ddpg/ddpg_evaluator.py

```diff
@@ -0,0 +1,167 @@
+        for var, target_var in zip(self.model.actor_var_list, self.target_model.actor_var_list):
+            a_updates.append(tf.assign(target_var, var))
+        actor_updates = tf.group(*a_updates)
+
+        c_updates = []
+        for var, target_var in zip(self.model.critic_var_list, self.target_model.critic_var_list):
+            c_updates.append(tf.assign(target_var, var))
+        critic_updates = tf.group(*c_updates)
+        self.sess.run([actor_updates, critic_updates])
+
+        self.sampler = SyncSampler(
+                        self.env, self.model, NoFilter(),
+                        config["num_local_steps"], horizon=config["horizon"])
+
+    def initialize(self):
+        self.sess.run(tf.global_variables_initializer())
+
+    def sample(self, no_replay = True):
+        """Returns a batch of samples."""
+
+        rollout = self.sampler.get_data()
+        rollout.data["weights"] = np.ones_like(rollout.data["rewards"])
+
+        samples = process_rollout(
+                    rollout, MeanStdFilter(()),
+                    gamma=self.config["gamma"], use_gae=False)
+
+        return samples
+
+    def _setup_target_updates(self):
+        """Set up actor and critic updates."""
+        a_updates = []
+        for var, target_var in zip(self.model.actor_var_list, self.target_model.actor_var_list):
+            a_updates.append(tf.assign(target_var,
+                    (1. - self.config["tau"]) * target_var
+                    + self.config["tau"] * var))
+        actor_updates = tf.group(*a_updates)
+
+        c_updates = []
+        for var, target_var in zip(self.model.critic_var_list, self.target_model.critic_var_list):
+            c_updates.append(tf.assign(target_var,
+                    (1. - self.config["tau"]) * target_var
+                    + self.config["tau"] * var))
+        critic_updates = tf.group(*c_updates)
+        self.target_updates = [actor_updates, critic_updates]
+
+    def update_target(self):
+        """Updates target critic and target actor."""
+        self.sess.run(self.target_updates)
+
+    def setup_gradients(self):
+        """Setups critic and actor gradients."""
+        self.critic_grads = tf.gradients(self.model.critic_loss, self.model.critic_var_list)
+        c_grads_and_vars = list(zip(self.critic_grads, self.model.critic_var_list))
+        c_opt = tf.train.AdamOptimizer(self.config["critic_lr"])
+        self._apply_c_gradients = c_opt.apply_gradients(c_grads_and_vars)
+
+        self.actor_grads = tf.gradients(self.model.actor_loss, self.model.actor_var_list)
+        a_grads_and_vars = list(zip(self.actor_grads, self.model.actor_var_list))
+        a_opt = tf.train.AdamOptimizer(self.config["actor_lr"])
+        self._apply_a_gradients = a_opt.apply_gradients(a_grads_and_vars)
+
+    def compute_gradients(self, samples):
+        """ Returns gradient w.r.t. samples."""
+        # actor gradients
+        actor_actions = self.sess.run(self.model.output_action,
```

Comment:
don't you track the actor samples in `samples["actions"]` or something?

Reply:
yeah, but looking at https://arxiv.org/pdf/1509.02971.pdf, aren't the actions according to what the actor currently outputs?

Reply:
ok yeah good point